File: 78-subsets/SolutionRecursiveAlternative.cc
class Solution {
public:
    void bktk(const vector<int>& a, int lev, vector<int>& cur, vector<vector<int>>& r){
        if(lev == a.size()){
            r.push_back(cur);
            return;
        }
        // without
        bktk(a, lev+1, cur, r);
        // with
        cur.push_back(a[lev]);
        bktk(a, lev+1, cur, r);
        cur.pop_back();
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> r;
        vector<int> cur;
        bktk(nums, 0, cur, r);
        return r;
    }
};

File: Leetcode/78-subsets/SolutionRecursiveAlternative.java
class Solution {
    void bktk(int[] nums, int lev, ArrayList<Integer> cur, List<List<Integer>> result){
        if(lev == nums.length){
            result.add(new ArrayList<Integer>(cur));
            return;
        }
            
        // without
        bktk(nums, lev+1, cur, result);
        // with
        cur.add(nums[lev]);
        bktk(nums, lev+1, cur, result);
        cur.remove(cur.size()-1);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        ArrayList<Integer> cur =  new ArrayList<Integer>();
        bktk(nums, 0, cur, result);
        return result;
    }
}